### DISCUSSION TITLE
Can I "listen" to changes in a `LiveObject`?

### DISCUSSION THREAD

**Discussion by prideout** [liveblocks/liveblocks#575] [Help] [answered]
I can't find anything in the docs that tells me how to subscribe to updates to a LiveObject.  I'm interested in knowing when any value in the LiveObject is modified, when any key is deleted, etc.

Maybe there's a way to do this using one of the provided middlewares?  I could use zustand's `subscribe` method I suppose, but it would give me the entire object; I'm interested in knowing exactly what portions have changed.

I'm aware of `Room.subscribe(storageItem)` but I would like to know what specific portions of the live structure were updated.

**Comment by nvie** [comment, marked as answer]
Hi @prideout, thanks for reaching out! Since you mentioned Zustand, I'm assuming you're using Zustand in your project? The idiomatic way to listen to changes in Zustand is to use the `useStorage()` hook and to use a selector for the data you're interested in if you want to rerender parts of your app based on when those values specifically change.

Here's an example:

```ts
useStore(state => state.my.data[0].field)
```

Using this in a component will rerender it whenever that nested field changes. Is this what you're trying to do?

If your problem isn't about rerendering your app based on changing data, but doing something else, you could indeed use the `Room.subscribe(storageItem)` API. In that case, you might want to use the `{ isDeep: true }` option in the subscription, which will also notify your callback when any of the nested data in there changes. The payload to these notifications is a list of updates that happened since the last change. Is this what you're looking for?

Here's an example:

```ts
room.subscribe(myLiveObject, (updates) => {
  for (const update of updates) {
    const {
      type,  // "update" | "delete"
      node,  // the subnode that changed
    } = update;

    // Do something here
  }
}, { isDeep: true });
```

**Reply by prideout** [reply]
Thank you @nvie the `isDeep` stuff fulfills my needs perfectly!  One point of confusion is that [the docs for subscribe](https://liveblocks.io/docs/api-reference/liveblocks-client#Room.subscribe(storageItem)) indicate that type is a string such as `LiveObject` or `LiveList`, but you've indicated that it is a string such as `update` or `delete`.

**Reply by nvie** [reply]
Oh, you're right, that _is_ confusing. This is now fixed — thanks!

**Reply by prideout** [reply]
Thanks @nvie !  If the user loses connectivity while my app is listening to a room, will it continue to receive notifications if another layer of the app is making changes to the room storage?

I am considering using live blocks to communicate between two local software layers, and I'm wondering if that would cause latency because of server round tripping.

**Reply by nvie** [reply]
Yeah, this should not be a problem. Local changes are applied optimistically, and don't depend on connectivity or a server. The update notifications will get sent anytime your storage node changes, whether it's due to an outside update (incoming via the network), or from local changes your app makes by updating storage.
